ipq807x: Add ZyXEL NBG7815

ddimension opened the pull request:
ZyXEL NBG7815 is a premium 802.11ax "tri"-band router/AP. 

Specifications:
* CPU: Qualcomm IPQ8072A Quad core Cortex-A53 2.2GHz
* RAM: 1 GB 2x Nanya NT5CC256M16ER-EK
* Storage:
 * 8MB serial flash Winbond W25Q64DW
 * 4GB eMMC flash Kingston EMMC04G-M627
* Ethernet:
 * 4x1G RJ45 ports (QCA8074A) with 1x status LED per port 
 * 1x2.5G RJ45 port (QCA8081) with 1x status LED 
 * 1x10G RJ45 port (AQR113C) with 1x status LED
* Switch: Qualcomm Atheros QCA8075
* WLAN:
* 2.4GHz: Qualcomm QCN5024 4x4@40MHz 802.11b/g/n/ax 1147 Mbps PHY rate
* 2x 5GHz: Qualcomm QCN5054 4x4 802.11a/b/g/n/ac/ax 2402 PHY rate
* Bluetooth CSR8811 using HSUART, currently unsupported
* USB: 1x USB3.0 Type-A port
* LED-s currently not supported: 
 * White
 * Dark Blu
 * Amber
 * Purple
 * Purple and dark blue 
 * Red
* Buttons:
* 1x Soft reset
* 1x WPS button
* Power: 12V DC Jack

Installation instructions:

See https://openwrt.org/inbox/toh/zyxel/nbg7815_armor_g5 for installation details.

Bluetooth usage:
* you need at least package bluez-utils, recommended bluez-daemon
* run following commands to enable and start 
   hciattach  /dev/ttyMSM1 bcsp
   hciconfig hci0 up

Many thanks to itorK for his work on this device:
https://github.com/itorK/openwrt/tree/nbg7815

Signed-off-by: André Valentin <avalentin@marcant.net>

marcant-oss committed:
ipq807x: Add ZyXEL NBG7815

ZyXEL NBG7815 is a premium 802.11ax "tri"-band router/AP.
Specifications:
* CPU: Qualcomm IPQ8072A Quad core Cortex-A53 2.2GHz
* RAM: 1 GB 2x Nanya NT5CC256M16ER-EK
* Storage:
        * 8MB serial flash Winbond W25Q64DW
        * 4GB eMMC flash Kingston EMMC04G-M627
* Ethernet:
        * 4x1G RJ45 ports (QCA8074A) with 1x status LED per port
        * 1x2.5G RJ45 port (QCA8081) with 1x status LED
        * 1x10G RJ45 port (AQR113C) with 1x status LED
* Switch: Qualcomm Atheros QCA8075
* WLAN:
        * 2.4GHz: Qualcomm QCN5024 4x4@40MHz 802.11b/g/n/ax 1147 Mbps PHY rate
        * 2x 5GHz: Qualcomm QCN5054 4x4 802.11a/b/g/n/ac/ax 2402 PHY rate
* Bluetooth CSR8811 using HSUART, currently unsupported
* USB: 1x USB3.0 Type-A port
* LED-s currently not supported:
        * White
        * Dark Blu
        * Amber
        * Purple
        * Purple and dark blue
        * Red
* Buttons:
        * 1x Soft reset
* Power: 12V DC Jack

Installation instructions:
* Disconnect WAN
* Reset device to factory defaults by pushing reset button 15 sec,
  LEDs should lit orange color.
* After 5-10 minutes, when the LEDs turn constant dark blue,
  put your LAN cable and connect at address 192.168.123.1 by telnet on port 23
* Login with
  NBG7815 login: root
  password: nbg7815@2019
* cd /tmp/ApplicationData
* wget -O openwrt-ipq807x-generic-zyxel_nbg7815-squashfs-sysupgrade.bin http://...
* wget https://github.com/itorK/nbg7815_tools/blob/main/flash_to_openwrt.sh
* run flash_to_openwrt.sh
If you can't use wget, you can transfer the files via nc.
See https://openwrt.org/inbox/toh/zyxel/nbg7815_armor_g5 for installation details.

Bluetooth usage:
* you need at least package bluez-utils, recommended bluez-daemon
* run following commands to enable and start
  hciattach  /dev/ttyMSM1 bcsp
  hciconfig hci0 up

Many thanks to itorK for his work on this device:
https://github.com/itorK/openwrt/tree/nbg7815

Reviewed-by: Robert Marko <robimarko@gmail.com>
Signed-off-by: André Valentin <avalentin@marcant.net>

ddimension commented:
@robimarko May I ask you to take a look?

Ansuel commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,607 @@
+			};
+		};
+
+	port_scheduler_resource {
```
please compare this with the ess node and drop it if it's the same.

Ansuel commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,607 @@
+	reset-gpios = <&tlmm 37 GPIO_ACTIVE_LOW>;
+
+	qca8075_1: ethernet-phy@0 {
+		reg = <0>;
```
please add `compatible = "ethernet-phy-ieee802.3-c22";` for each c22 generic node

Ansuel commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,607 @@
+};
+
+&soc {
+	dp1: dp1 {
+		device_type = "network";
+		compatible = "qcom,nss-dp";
+		qcom,id = <1>;
+		reg = <0x3a001000 0x200>;
+		qcom,mactype = <0>;
+		local-mac-address = [00 00 00 00 00 00];
+		phy-handle = <&qca8075_1>;
+		phy-mode = "sgmii";
+		label = "lan1";
+	};
```
Please drop the redundant part... check other dts on how this is done...
in theory only the phy-handle and label are needed

robimarko replied:
I second this, you can use the existing labels from ESS DTSI

Ansuel commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,607 @@
+		label = "wan";
+	};
+
+	dp6: dp6 {
```
use the syn variant for mactype 1 dp6 node

Ansuel commented on target/linux/ipq807x/base-files/lib/preinit/10_fix_eth_mac.sh:
```diff
@@ -0,0 +1,16 @@
+		ip link set dev lan3 address $ethaddr
+		ip link set dev lan4 address $ethaddr
+		ip link set dev 10g address $ethaddr
+		ip link set dev wan address $(macaddr_add $ethaddr 1)
```
why this is needed? mac are not assigned by uboot and copied to the local-mac-address binding?

robimarko replied:
Even if not set, it should be possible to use NVMEM-s U-boot driver to read the cell directly in DTS

ddimension replied:
> Even if not set, it should be possible to use NVMEM-s U-boot driver to read the cell directly in DTS

I tried last weekend and it was not working. I do not know if it was the dynamically declared u-boot env partition or missing nvmem support in the network driver.

ddimension replied:
> why this is needed? mac are not assigned by uboot and copied to the local-mac-address binding?

There only one mac address in u-boot env. That's why it only works for eth0 / lan1.

Ansuel replied:
afaik the uboot nvmem driver still require some love to correctly work... for now i would use the uboot way that is probably supported

ddimension replied:
> > Even if not set, it should be possible to use NVMEM-s U-boot driver to read the cell directly in DTS
> 
> I tried last weekend and it was not working. I do not know if it was the dynamically declared u-boot env partition or missing nvmem support in the network driver.

Just that that you fixed it by specifying the partitions manually. Will do that.

ddimension replied:
Cloud

> Even if not set, it should be possible to use NVMEM-s U-boot driver to read the cell directly in DTS

Could you take a look at the dts? I've changed it, added the kernel symbol for nvmem uboot-env. But always random macs...

ddimension replied:
That's the error in dmesg:
OF: /soc/dp1: could not get #nvmem-cell-cells for /soc/spi@78b5000/flash@0/partitions/partition@600000/ethaddr

Ansuel replied:
in theory should be just a wanr to ignore. Does the mac gets correctly assigned?

Ansuel replied:
@ddimension can you give output of fw_printenv ?

musashino205 replied:
Please try to use ```mac-address-ascii``` instead of ```mac-address``` in ```nvmem-cell-names```. Currently, Linux Kernel doesn't support ascii-based MAC address via NVMEM and the downstream implementation of ```of_get_mac_address``` is necessary.

example: https://github.com/openwrt/openwrt/pull/11228/files?diff=unified&w=0#diff-1fd579c71484b41770913dd23ce55bbb7f73dc034a8636d48239154d3c770ccbR209

robimarko replied:
That NVMEM warning is false and there is a patch pending to fix it:
https://lore.kernel.org/lkml/DU0PR04MB941732E9BB61E51ADD66F82088FC9@DU0PR04MB9417.eurprd04.prod.outlook.com/T/

Also, Rafal has been working on adding ASCII support to NVMEM

ddimension replied:
Yepp, it works :-)

robimarko commented on target/linux/ipq807x/base-files/lib/upgrade/platform.sh:
```diff
@@ -66,6 +66,18 @@ platform_do_upgrade() {
 		mmc_do_upgrade "$1"
 		;;
+	zyxel,nbg7815)
+		part_num="$(hexdump -e '1/1 "%01x|"' -n 1 -s 168 -C /dev/mtd2|cut -f 1 -d "|"|head -n1)"
```
Please no partition hardcoding, there are generic helpers to find the mtd index by partition name instead

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,607 @@
+		serial0 = &blsp1_uart5;
+		serial1 = &blsp1_uart3;
+		/* Alias as required by u-boot to patch MAC addresses */
+		ethernet0 = &dp1;
```
This ties to the preinit question, you sure it only knows one address?

ddimension replied:
Absolutely. Checked the env!

robimarko replied:
Since NVMEM is now used, there is no use for this alias, remove it and the comment and just leave the label MAC alias

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,607 @@
+			bias-disable;
+	};
+
+	button_pins {
```
This and `gpio_pinmux` are completely unused

ddimension replied:
Okay, will remove it. Thanks!

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,607 @@
+	};
+
+	aqr113c_5: ethernet-phy@5 {
+		compatible ="ethernet-phy-ieee802.3-c45";
```
Missing space, and you dont need the `_5` in the label as its the only AQR113C

robimarko commented:
Please put the install instructions directly in the commit itself as they are not that long.
Also, if bluetooth is working, why not include the required packages by default?

ddimension commented:
> Please put the install instructions directly in the commit itself as they are not that long. Also, if bluetooth is working, why not include the required packages by default?

I could kmod-bluetooth. But the whole bluez stuff is not in the openwrt repo itself, so it's kind of useless. 
Also, it is needed to run <code>hciattach  /dev/ttyMSM1 bcsp</code> to enable start it, which requires openwrt-packages repo.

marcant-oss committed:
ipq807x: add nvmem uboot-env support

Enabling kernel symbol CONFIG_NVMEM_U_BOOT_ENV allows to use u-boot
environement variable ethaddr with nvmen. That way it is possible to assign
the MAC address to the ethernet device driver.

Example of usage in dts:
....
	partition@600000 {
			compatible = "u-boot,env";
			label = "0:appsblenv";
			reg = <0x600000 0x10000>;

			macaddr_lan: ethaddr {
			};
	};
....
&dp5 {
	status = "okay";
	phy-handle = <&qca8081>;
	label = "wan";
	nvmem-cells = <&macaddr_lan>;
	nvmem-cell-names = "mac-address-ascii";
	mac-address-increment = <1>;
};

This is needed for Zyxel NBG7815.

Signed-off-by: André Valentin <avalentin@marcant.net>

Ansuel commented on target/linux/ipq807x/base-files/lib/upgrade/platform.sh:
```diff
@@ -66,6 +66,20 @@ platform_do_upgrade() {
 		mmc_do_upgrade "$1"
 		;;
+	zyxel,nbg7815)
+		local config_mtdnum="$( find_mtd_index 0:bootconfig )"
```
nitpick no space at start and end of ( )

Ansuel commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,444 @@
+		port@0 {
+			port_id = <1>;
+			phy_address = <0>;
+		};
+		port@1 {
```
nitpick space between nodes

Ansuel commented:
Anyway looks good to me... We just need to understand how to handle the mac address...

robimarko commented:
If `mac-address-ascii` works then:
`Reviewed-by: Robert Marko <robimarko@gmail.com>`

ddimension commented:
> Please put the install instructions directly in the commit itself as they are not that long. Also, if bluetooth is working, why not include the required packages by default?

I've added the instructions and added the script to the openwrt wiki.

Ansuel commented:
@ddimension i miss did we sort the problem with mac address?

ddimension commented:
> @ddimension i miss did we sort the problem with mac address?

Yepp, It has been fixed with the u-boot-env nvmem provider :-) No more scripting.
Really cool stuff!

Ansuel reviewed (changes requested):
Can you set a better commit description for this? Like why is needed also for other target? (quote the use of uboot env)

(the commit adding the uboot nvmem config)

ddimension commented:
> Can you set a better commit description for this? Like why is needed also for other target? (quote the use of uboot env)
> 
> (the commit adding the uboot nvmem config)

Please notice I only enabled it for ipq807x, not for all targets.

Ansuel commented:
I know it's just to have something more descriptive than "It's for x"

ddimension commented:
> I know it's just to have something more descriptive than "It's for x"

Of course,  would have been easier if I had such a reference.

Ansuel commented:
@robimarko think he can add your review tag?

robimarko commented:
I already did above:
https://github.com/openwrt/openwrt/pull/11818#issuecomment-1386975776

Ansuel commented:
Yep but since things changed... anyway was a hint to add you tag to the commit :D

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,449 @@
+		};
+	};
+
+	i2c_1_pins: i2c_1_pinmux {
```
This is not used at all currently

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8074-nbg7815.dts:
```diff
@@ -0,0 +1,449 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/* Copyright (c) 2022, Karol Przybylski <itor@o2.pl> */
```
Since there will be changes, bump to 2023

ddimension replied:
Good catch, also updates the person list!

robimarko commented:
Good thing you reminded me, I looked at the DTS again and posted some comments.

After those are resolved:
`Reviewed-by: Robert Marko <robimarko@gmail.com>`

ddimension commented:
> Good thing you reminded me, I looked at the DTS again and posted some comments.
> 
> After those are resolved: `Reviewed-by: Robert Marko <robimarko@gmail.com>`

Solved all issues!

ddimension commented:
Thank you for this really fast merge and review!

hnyman commented:
Out of curiosity:

* Does the factory image need to be compiled  at all? To me it looks like only sysupgrade image is actually used.

* Could the "flash script" be packaged as a third item in the sysupgrade.tar, next to kernel and rootfs. Then it would not need to be downloaded from itorK's github site. (but the user would need to extract the tar file)
* Alternatively, offer the flash script as a downloadable item from the ipq807x download directory.
